File: node/zkgroup/receipts/ClientZkReceiptOperations.ts
/*
 *
 * Copyright (C) 2021 Signal Messenger, LLC.
 * All rights reserved.
 *
 * SPDX-License-Identifier: GPL-3.0-only
 *
 */

import {randomBytes} from 'crypto';
import FFICompatArray, {FFICompatArrayType} from '../internal/FFICompatArray';
import ZkGroupError from '../errors/ZkGroupError';
import {RANDOM_LENGTH} from '../internal/Constants';
import Native, {FFI_RETURN_OK} from '../internal/Native';
import ServerPublicParams from '../ServerPublicParams';
import ReceiptCredential from "./ReceiptCredential";
import ReceiptCredentialPresentation from "./ReceiptCredentialPresentation";
import ReceiptCredentialRequestContext from "./ReceiptCredentialRequestContext";
import ReceiptCredentialResponse from "./ReceiptCredentialResponse";
import ReceiptSerial from "./ReceiptSerial";

export default class ClientZkReceiptOperations {

    serverPublicParams: ServerPublicParams

    constructor(serverPublicParams: ServerPublicParams) {
        this.serverPublicParams = serverPublicParams;
    }

    createReceiptCredentialRequestContext(receiptSerial: ReceiptSerial): ReceiptCredentialRequestContext {
        const random = new FFICompatArray(randomBytes(RANDOM_LENGTH));
        return this.createReceiptCredentialRequestContextWithRandom(random, receiptSerial);
    }

    createReceiptCredentialRequestContextWithRandom(random: FFICompatArrayType, receiptSerial: ReceiptSerial): ReceiptCredentialRequestContext {
        const newContents = new FFICompatArray(ReceiptCredentialRequestContext.SIZE);
        const serverPublicParamsContents = this.serverPublicParams.getContents();
        const receiptSerialContents = receiptSerial.getContents();

        const ffi_return = Native.FFI_ServerPublicParams_createReceiptCredentialRequestContextDeterministic(
            serverPublicParamsContents, serverPublicParamsContents.length,
            random, random.length,
            receiptSerialContents, receiptSerialContents.length,
            newContents, newContents.length);

        if (ffi_return != FFI_RETURN_OK) {
            throw new ZkGroupError("FFI_RETURN!=OK");
        }

        return new ReceiptCredentialRequestContext(newContents);
    }

    receiveReceiptCredential(receiptCredentialRequestContext: ReceiptCredentialRequestContext, receiptCredentialResponse: ReceiptCredentialResponse): ReceiptCredential {
        const newContents = new FFICompatArray(ReceiptCredential.SIZE);
        const serverPublicParamsContents = this.serverPublicParams.getContents();
        const receiptCredentialRequestContextContents = receiptCredentialRequestContext.getContents();
        const receiptCredentialResponseContents = receiptCredentialResponse.getContents();

        const ffi_return = Native.FFI_ServerPublicParams_receiveReceiptCredential(
            serverPublicParamsContents, serverPublicParamsContents.length,
            receiptCredentialRequestContextContents, receiptCredentialRequestContextContents.length,
            receiptCredentialResponseContents, receiptCredentialResponseContents.length,
            newContents, newContents.length);

        if (ffi_return != FFI_RETURN_OK) {
            throw new ZkGroupError("FFI_RETURN!=OK");
        }

        return new ReceiptCredential(newContents);
    }

    createReceiptCredentialPresentation(receiptCredential: ReceiptCredential): ReceiptCredentialPresentation {
        const random = new FFICompatArray(randomBytes(RANDOM_LENGTH));
        return this.createReceiptCredentialPresentationWithRandom(random, receiptCredential);
    }

    createReceiptCredentialPresentationWithRandom(random: FFICompatArrayType, receiptCredential: ReceiptCredential): ReceiptCredentialPresentation {
        const newContents = new FFICompatArray(ReceiptCredentialPresentation.SIZE);
        const serverPublicParamsContents = this.serverPublicParams.getContents();
        const receiptCredentialContents = receiptCredential.getContents();

        const ffi_return = Native.FFI_ServerPublicParams_createReceiptCredentialPresentationDeterministic(
            serverPublicParamsContents, serverPublicParamsContents.length,
            random, random.length,
            receiptCredentialContents, receiptCredentialContents.length,
            newContents, newContents.length);

        if (ffi_return != FFI_RETURN_OK) {
            throw new ZkGroupError("FFI_RETURN!=OK");
        }

        return new ReceiptCredentialPresentation(newContents);
    }
}

File: java/java/src/main/java/org/signal/zkgroup/receipts/ClientZkReceiptOperations.java
//
// Copyright 2020-2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.zkgroup.receipts;

import java.security.SecureRandom;
import org.signal.zkgroup.InvalidInputException;
import org.signal.zkgroup.ServerPublicParams;
import org.signal.zkgroup.VerificationFailedException;
import org.signal.client.internal.Native;

import static org.signal.zkgroup.internal.Constants.RANDOM_LENGTH;

public class ClientZkReceiptOperations {

  private final ServerPublicParams serverPublicParams;

  public ClientZkReceiptOperations(ServerPublicParams serverPublicParams) {
    this.serverPublicParams = serverPublicParams;
  }

  public ReceiptCredentialRequestContext createReceiptCredentialRequestContext(ReceiptSerial receiptSerial) throws VerificationFailedException {
    return createReceiptCredentialRequestContext(new SecureRandom(), receiptSerial);
  }

  public ReceiptCredentialRequestContext createReceiptCredentialRequestContext(SecureRandom secureRandom, ReceiptSerial receiptSerial) throws VerificationFailedException {
    byte[] random      = new byte[RANDOM_LENGTH];
    secureRandom.nextBytes(random);

    byte[] newContents = Native.ServerPublicParams_CreateReceiptCredentialRequestContextDeterministic(serverPublicParams.getInternalContentsForJNI(), random, receiptSerial.getInternalContentsForJNI());

    try {
      return new ReceiptCredentialRequestContext(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

  public ReceiptCredential receiveReceiptCredential(ReceiptCredentialRequestContext receiptCredentialRequestContext, ReceiptCredentialResponse receiptCredentialResponse) throws VerificationFailedException {
    byte[] newContents = Native.ServerPublicParams_ReceiveReceiptCredential(serverPublicParams.getInternalContentsForJNI(), receiptCredentialRequestContext.getInternalContentsForJNI(), receiptCredentialResponse.getInternalContentsForJNI());

    try {
      return new ReceiptCredential(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

  public ReceiptCredentialPresentation createReceiptCredentialPresentation(ReceiptCredential receiptCredential) throws VerificationFailedException {
    return createReceiptCredentialPresentation(new SecureRandom(), receiptCredential);
  }

  public ReceiptCredentialPresentation createReceiptCredentialPresentation(SecureRandom secureRandom, ReceiptCredential receiptCredential) throws VerificationFailedException {
    byte[] random      = new byte[RANDOM_LENGTH];
    secureRandom.nextBytes(random);

    byte[] newContents = Native.ServerPublicParams_CreateReceiptCredentialPresentationDeterministic(serverPublicParams.getInternalContentsForJNI(), random, receiptCredential.getInternalContentsForJNI());

    try {
      return new ReceiptCredentialPresentation(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

}
